routes/dao_coin_exchange_with_fees.go

```diff
@@ -0,0 +1,1602 @@
+func GetQuoteBasePkidFromBuyingSellingPkids(
+	buyingPkid string,
+	sellingPkid string,
+	side string,
+) (
+	_quoteCurrencyPkid string,
+	_baseCurrencyPkid string,
+	_err error,
+) {
+	// The rule of thumb is we're seeling the base with an ask and buying the
+	// base with a bid.
+	if side == lib.DAOCoinLimitOrderOperationTypeBID.String() {
+		return sellingPkid, buyingPkid, nil
+	} else if side == lib.DAOCoinLimitOrderOperationTypeASK.String() {
+		return buyingPkid, sellingPkid, nil
+	} else {
+		return "", "", fmt.Errorf(
+			"GetQuoteBasePkidFromBuyingSellingPkids: Invalid side: %v", side)
+	}
+}
+
+type GetQuoteCurrencyPriceInUsdRequest struct {
+	QuoteCurrencyPublicKeyBase58Check string `safeForLogging:"true"`
+}
+
+type GetQuoteCurrencyPriceInUsdResponse struct {
+	UsdPrice string `safeForLogging:"true"`
+}
+
+func (fes *APIServer) GetQuoteCurrencyPriceInUsdEndpoint(ww http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := GetQuoteCurrencyPriceInUsdRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetQuoteCurrencyPriceInUsd: Problem parsing request body: %v", err))
+		return
+	}
+
+	usdPrice, err := fes.GetQuoteCurrencyPriceInUsd(requestData.QuoteCurrencyPublicKeyBase58Check)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetQuoteCurrencyPriceInUsd: Problem getting quote currency price in USD: %v", err))
+		return
+	}
+
+	res := GetQuoteCurrencyPriceInUsdResponse{UsdPrice: usdPrice}
+	if err := json.NewEncoder(ww).Encode(res); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetQuoteCurrencyPriceInUsd: Problem encoding response: %v", err))
+		return
+	}
+}
+
+func (fes *APIServer) GetQuoteCurrencyPriceInUsd(
+	quoteCurrencyPublicKey string) (string, error) {
+	if IsDesoPkid(quoteCurrencyPublicKey) {
+		desoUsdCents := fes.GetExchangeDeSoPrice()
+		return fmt.Sprintf("%0.9f", float64(desoUsdCents)/100), nil
+	}
+	utxoView, err := fes.backendServer.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Error fetching mempool view: %v", err)
+	}
+
+	pkBytes, _, err := lib.Base58CheckDecode(quoteCurrencyPublicKey)
+	if err != nil || len(pkBytes) != btcec.PubKeyBytesLenCompressed {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Problem decoding public key %s: %v",
+			quoteCurrencyPublicKey, err)
+	}
+
+	existingProfileEntry := utxoView.GetProfileEntryForPublicKey(pkBytes)
+	if existingProfileEntry == nil || existingProfileEntry.IsDeleted() {
+		return "", fmt.Errorf(
+			"GetQuoteCurrencyPriceInUsd: Profile for quote currency public "+
+				"key %v does not exist",
+			quoteCurrencyPublicKey)
+	}
+
+	// If the profile is the dusdc profile then just return 1.0
+	lowerUsername := strings.ToLower(string(existingProfileEntry.Username))
+	if strings.Contains(lowerUsername, "dusdc") {
```

Comment:
Just checking the username is insufficient. someone could make a new username that includes `dusdc` and could lead to unexpected results. Can we add a validation to ensure we have approved stuff coming in for `quoteCurrencyPublicKey`?

Reply:
Ah you're right! I just changed it to an exact match, which I believe should be robust to this. I think before I wasn't sure if we'd have the same setup on devnet/testnet/mainnet, but we do so it's fine.